### PR TITLE
Test connecting tenants via password over TLS through proxy

### PR DESCRIPTION
**Pull request by tbg**
This commit adds infrastructure for testing secure connections through
the multi-tenancy proxy to a standalone SQL server. This works "out of
the box" for essentially those tests for which we previously had client
certificates working.

The tests that fail here are GoPG, Hibernate, Sequelize, Django and they
all fail for the basic reason that they don't support secure query
strings, at least not in the way they are supplied right now. To
*really* find out whether they have issues with multi-tenancy (or even
client certs), we need to customize the way they receive the connection
parameters. This is not something I plan to investigate.

Closes https://github.com/cockroachdb/cockroach/issues/52413.

### CODE REVIEW

**Comment by tbg** [comment]
Friendly ping.

File: go.mod [hunk 1]
```diff
@@ -3,7 +3,7 @@ module github.com/cockroachdb/examples-orms
 go 1.13
 
 require (
-	github.com/cockroachdb/cockroach-go/v2 v2.0.6
+	github.com/cockroachdb/cockroach-go/v2 v2.0.7-0.20200903114400-4bf206d91533
```
Comment (rafiss) [inline go.mod:6]: Why can't this use a normal tag?

File: testing/main_test.go [hunk 2]
```diff
@@ -368,66 +403,111 @@ FROM
...
+			authPassword: "password via query string unsupported",
+		},
+	)
 }
 
+// TODO(during review): why can't I run the ActiveRecords tests manually
+// with this invocation?
+//
+//   ./docker.sh make deps
+//   ./docker.sh go test -v -run ActiveRecord ./testing
```
Comment (rafiss) [inline testing/main_test.go:498]: Weird.. mine also fails locally, but I get an error:

```
    main_test.go:296: Get "http://localhost:6543/ping/": dial tcp 127.0.0.1:6543: connect: connection refused
=== RUN   TestActiveRecord/SystemTenant/SecondRun
make: Entering directory '/home/rafiss/go/src/github.com/cockroachdb/examples-orms/ruby/activerecord'
panic: test timed out after 10m0s

rake aborted!
PG::ConnectionBad: could not connect to server: Connection refused
	Is the server running on host "localhost" (127.0.0.1) and accepting
	TCP/IP connections on port 34497?
could not connect to server: Cannot assign requested address
	Is the server running on host "localhost" (::1) and accepting
	TCP/IP connections on port 34497?
```

File: go.mod (hunk 1 above)
Comment (tbg) [reply go.mod:6]: It does now. When I opened the PR, the PR in cockroach-go hadn't merged yet.

**Comment by tbg** [comment]
PTAL

**Review by rafiss — APPROVED** [review]
LGTM, though perhaps we could change the `TODO(during review)` to a different phrasing. I'd like to keep your notes on the issue there, but just rephrased so it doesn't seem like the comment fell through the cracks during a review.

**Comment by tbg** [comment]
> LGTM, though perhaps we could change the TODO(during review) to a different phrasing. I'd like to keep your notes on the issue there, but just rephrased so it doesn't seem like the comment fell through the cracks during a review.

Oh, yep, that was an oversight! Changed to a `TODO(rafiss)`.

**Review by rafiss — APPROVED** [review]
thanks for the work here!

**Comment by tbg** [comment]
Uh, the build has been running for 20h! I only changed a comment. Going to restart it and :crossed_fingers: but is this something you've seen before? It's sitting in

```
[12:36:05]
[TestActiveRecord/insecure/RegularTenant/SecondRun] TestActiveRecord/insecure/RegularTenant/SecondRun
[12:36:05]
[TestActiveRecord/insecure/RegularTenant/SecondRun] TestActiveRecord/insecure/RegularTenant/SecondRun/RetrieveFromAPIAfterRestart (running for 20h:38m:04s)
[12:36:05]
[TestActiveRecord/insecure/RegularTenant/SecondRun/RetrieveFromAPIAfterRestart] TestActiveRecord/insecure/RegularTenant/SecondRun/RetrieveFromAPIAfterRestart
[12:36:05]
[TestActiveRecord/insecure/RegularTenant/SecondRun/RetrieveFromAPIAfterRestart] TestActiveRecord/insecure/RegularTenant/SecondRun/RetrieveFromAPIAfterRestart/Customers
[12:36:05]
[TestActiveRecord/insecure/RegularTenant/SecondRun/RetrieveFromAPIAfterRestart] TestActiveRecord/insecure/RegularTenant/SecondRun/RetrieveFromAPIAfterRestart/Products
[12:36:05]
[TestActiveRecord/insecure/RegularTenant/SecondRun/RetrieveFromAPIAfterRestart] TestActiveRecord/insecure/RegularTenant/SecondRun/RetrieveFromAPIAfterRestart/Order
```

**Comment by tbg** [comment]
Hmm, hanging again. These tests pass on master right now, right?

**Comment by rafiss** [comment]
Yikes... yes they are passing on master right now. So either (1) CockroachDB master has changed so that it no longer works with your PR, or (2) the versions of the libraries used by this PR have changed (e.g. always fetching latest) and the new versions don't work.

We have seen (2) happen before but AFAIK we have pinned all versions by now.

**Comment by rafiss** [comment]
Here is an older passing test run for this PR: https://teamcity.cockroachdb.com/viewLog.html?buildId=2259861&buildTypeId=Cockroach_ExampleORMsUnitTests&tab=buildLog&branch_Cockroach=%3Cdefault%3E&_focus=325#_state=325

The hanging test is ActiveRecord. Both that previously passing test and the currently hanging test are using `activerecord 5.2.4.3` and `activerecord-cockroachdb-adapter 5.2.0` (and all other transitive dependencies are the same).

**Comment by tbg** [comment]
Hmm, this is annoying.

I've decided to abandon this PR for the time being. The hope that we would get some real coverage for multi tenancy from this hasn't materialized, due to most ORMs here not being configurable with SSL as it stands. Instead, @darinpp has tested various ORMs manually and I don't think he's going to lean on this PR at this point.

Sorry to have wasted your time here!

**Comment by rafiss** [comment]
@tbg  No problem!

Just a point of clarification to make sure I'm on the same page. Is this the testing you are referring to? https://docs.google.com/document/d/1cpqBG4fpMiytFMY2s7MD0l7iTamxUnoKok-EhvD9_WY/edit

That is mostly driver testing, not ORM testing. So my outsider opinion is that it would still be nice to test the user experience with ORMs, if possible.

**Comment by tbg** [comment]
👍 thanks for the clarification. cc @darinpp - if you (or someone on the CC end) wanted to pick this up, it's all yours.

**Pull request closed without merging**